aea/manager/project.py

```diff
@@ -172,17 +173,28 @@ def _ensure_private_keys(self) -> None:
         """Add private keys if not present in the config."""
         builder = self._get_builder(self.agent_config, self.project.path)
         default_ledger = builder.get_default_ledger()
+        required_ledgers = builder.get_required_ledgers()
+        enforce(
+            default_ledger in required_ledgers,
+            exception_text=f"Default ledger '{default_ledger}' not in required ledgers: {required_ledgers}",
+            exception_class=AEAValidationError,
+        )
 
-        if not self.agent_config.private_key_paths.read_all():
-            self.agent_config.private_key_paths.create(
-                default_ledger, self._create_private_key(default_ledger)
-            )
+        available_private_keys = self.agent_config.private_key_paths.keys()
+        available_connection_private_keys = (
+            self.agent_config.connection_private_key_paths.keys()
+        )
 
-        if not self.agent_config.connection_private_key_paths.read_all():
-            self.agent_config.connection_private_key_paths.create(
-                default_ledger,
-                self._create_private_key(default_ledger, is_connection=True),
-            )
+        for required_ledger in set(required_ledgers):
+            if required_ledger not in available_private_keys:
+                self.agent_config.private_key_paths.create(
+                    required_ledger, self._create_private_key(required_ledger)
+                )
+            if required_ledger not in available_connection_private_keys:
+                self.agent_config.connection_private_key_paths.create(
+                    required_ledger,
+                    self._create_private_key(required_ledger, is_connection=True),
+                )
```

Comment:
for each required ledger identifier, if there isn't a key of that type already, we generate one for 'private keys' and one for 'connection'

Reply:
the check that the default ledger is in required ledgers is done at line 176